Client/Frontend/Browser/FaviconManager.swift

```diff
@@ -57,43 +57,44 @@ class FaviconManager: TabHelper {
-                                            date: Date(),
-                                            type: icon.type)
-                                        
-                                        guard let img = img else {
-                                            deferred.fill(Maybe(failure: FaviconError()))
-                                            return
-                                        }
-                                        fav.width = Int(img.size.width)
-                                        fav.height = Int(img.size.height)
-                                        
-                                        if !tab.isPrivate {
-                                            if tab.favicons.isEmpty {
-                                                self.makeFaviconAvailable(tab, atURL: currentURL, favicon: fav, withImage: img)
-                                            }
-                                            tab.favicons.append(fav)
-                                            self.profile.favicons.addFavicon(fav, forSite: site).upon { _ in
-                                                deferred.fill(Maybe(success: fav))
-                                            }
-                                        } else {
-                                            tab.favicons.append(fav)
-                                            deferred.fill(Maybe(success: fav))
-                                        }
-        })
+
+        func loadImageCompleted(_ img: UIImage?, _ url: URL?) {
+            guard let img = img, let urlString = url?.absoluteString else {
+                deferred.fill(Maybe(failure: FaviconError()))
+                return
+            }
+
+            let fav = Favicon(url: urlString, date: Date(), type: icon.type)
+            fav.width = Int(img.size.width)
+            fav.height = Int(img.size.height)
+
+            if !tab.isPrivate {
+                if tab.favicons.isEmpty {
+                    self.makeFaviconAvailable(tab, atURL: currentURL, favicon: fav, withImage: img)
+                }
+                tab.favicons.append(fav)
+                self.profile.favicons.addFavicon(fav, forSite: site).upon { _ in
+                    deferred.fill(Maybe(success: fav))
+                }
+            } else {
+                tab.favicons.append(fav)
+                deferred.fill(Maybe(success: fav))
+            }
+        }
+
+        var fetch: SDWebImageOperation? = nil
+        fetch = manager.loadImage(with: iconUrl, options: SDWebImageOptions(options),
+                                  progress: { (receivedSize, expectedSize, _) in
+                                    if receivedSize > FaviconManager.maximumFaviconSize || expectedSize > FaviconManager.maximumFaviconSize {
+                                        fetch?.cancel()
+                                    }
+                                  })
+                                {  (img, _, _, _, _, url) in
```

Comment:
Nit: with multiple closure args, I wonder if it's not actually clearer to keep the named argument and these within the parens.

Reply:
The sheer number of unused named return params increased my cognitive load when updating the PR, so this seemed lower-noise when going through this code. I am not sure what the preferred style is.

Reply:
Ah, I think we're talking about two different things. 

I _love_ the underscores.

What I'm talking about is: 


```swift
object.method(withProgress: { progress in }) { success in }
```
versus
```swift
object.method(withProgress: { progress in }, andCompletion: { success in })
```

With method calls with two closures as parameters, I feel we should be a bit more careful with the last-closure-shorthand.

Reply:
Oh ok, I hadn't thought about that one. I see your point, it does look weird with the other closure as a named param.
I have (yet) another update to this patch because of how we compare UIImages in testing is failing, so I'll change these to a named param.